src/Language/Java/Inline.hs

```diff
@@ -394,3 +396,8 @@ blockQQ input = case Java.parser Java.block input of
       -- returning boxed values. Once this limitation of the compiler gets
       -- lifted, we'll support returning unboxed values, just like `call` does.
       castReturnType funcall = [| unsafeUncoerce . coerce <$> $funcall |]
+      unfoldHeadTySyn ty@(TH.ConT nT) = do
+        TH.reify nT >>= \case
+          TH.TyConI (TH.TySynD _ _ ty') -> unfoldHeadTySyn ty'
+          _ -> return ty
```

Comment:
Yes. Once you have the head, if it's not a `J ty` then some suitable `Coercible` instance will be selected.